### ISSUE TITLE
CPU and GPU trainings give different results

### ISSUE THREAD

**Issue by qobilidop** [qobilidop/srcnn#6]
We get drastically different results when using CPU or GPU for training. If we look at the PSNR metric, a CPU trained model oscillates around 27.

![history_psnr](https://cloud.githubusercontent.com/assets/5176695/23924547/4e4837fa-08c8-11e7-9d30-1d678a6d6bb5.png)

If we use GPU, PSNR goes up to around 33. Also strange here is that there are many `inf` PSNR values in this case, which are the missing points in the plot.

![history_psnr](https://cloud.githubusercontent.com/assets/5176695/23924552/58a2c53a-08c8-11e7-9107-60a7c27e447b.png)

What could have caused this CPU/GPU difference?

**Comment by qobilidop** [comment]
I learned that GPU reductions are not deterministic for the sake of performance, but I'm not sure if that solves this issue. See the following discussions:

* [Mention that GPU reductions are nondeterministic in docs](https://github.com/tensorflow/tensorflow/issues/2732)
* [Non-deterministic mean and sum reduction](https://github.com/tensorflow/tensorflow/issues/3103)
* [Determinism in tensorflow gradient updates?](http://stackoverflow.com/questions/39938307/determinism-in-tensorflow-gradient-updates)

**Comment by JasonAn** [comment]
I repeated the training "the train_ep250.py"  on my desktop. The psnr range is 30-35 and equals inf sometimes.

Changing to ssim as the metric, it reduces the inf and runs much smoother since the two constants C1, and C2  in the definition.

**Comment by qobilidop** [comment]
@JasonAn It's good to know that. So I guess the difference is caused by different code versions. @jiacong1990 got the CPU results last night, after which I might have changed the weights initialization.

**Comment by qobilidop** [comment]
I checked the config files. Indeed the initialization for the last layer changed from `glorot_uniform` to `he_normal`. So we learn from this that initialization matters a lot and `he_normal` works great.

Also Keras 2.0.0 released today, and our config file format changed a little.

**Comment by qobilidop** [comment]
The `inf` PSNR issue is still there.
